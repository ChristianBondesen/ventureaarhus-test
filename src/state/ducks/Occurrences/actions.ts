import { IFluxStandardAction } from '../FluxStandardAction';
import * as types from './types';

// Interface for your actions -->
export interface ISetOccurrences extends IFluxStandardAction {
  type: types.SET_OCCURRENCES;
  payload: types.IOccurrence[];
}

export interface ISetOccurrencesTags extends IFluxStandardAction {
  type: types.SET_OCCURRENCES_TAGS;
  payload: types.ITags[];
}

export interface IUpdateRecommendedOccurrences extends IFluxStandardAction {
  type: types.UPDATE_RECOMMENDED_OCCURRENCES;
  payload: types.IOccurrence[];
}

// State of your Actions -->
// (Import for the reducer)
export type OccurrenceStateAction =
  | ISetOccurrences
  | ISetOccurrencesTags
  | IUpdateRecommendedOccurrences;

// Actual action implementing the former defined interface -->
export const setOccurrenceList = (
  occurrences: types.IOccurrence[]
): ISetOccurrences => ({
  type: types.SET_OCCURRENCES,
  payload: occurrences,
});

export const setOccurrencesTags = (
  tags: types.ITags[]
): ISetOccurrencesTags => ({
  type: types.SET_OCCURRENCES_TAGS,
  payload: tags,
});

export const updateRecommendedOccurrences = (
  recOccurs: types.IOccurrence[]
): IUpdateRecommendedOccurrences => ({
  type: types.UPDATE_RECOMMENDED_OCCURRENCES,
  payload: recOccurs,
});

// Interface of all your Actions -->
export interface IOccurrenceActions {
  setOccurrenceList(occurrences: types.IOccurrence[]);
  setOccurrencesTags(tags: types.ITags[]);
  updateRecommendedOccurrences(recOccurs: types.IOccurrence[]);
}

// Export the actions -->
export default {
  setOccurrenceList,
  setOccurrencesTags,
  updateRecommendedOccurrences,
};
