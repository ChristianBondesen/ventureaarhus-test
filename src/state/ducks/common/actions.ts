import { IFluxStandardAction } from '../FluxStandardAction';
import * as types from './types';

export interface IChangeConnectionStatus extends IFluxStandardAction {
  type: types.CHANGE_CONNECTION_STATUS;
  payload: boolean;
}
export interface IIncrementNetworkCalls extends IFluxStandardAction {
  type: types.INCREMENT_NETWORK_CALLS;
}

export type CommonStateAction =
  | IChangeConnectionStatus
  | IIncrementNetworkCalls;

export const setConnectionStatus = (
  connected: boolean
): IChangeConnectionStatus => ({
  type: types.CHANGE_CONNECTION_STATUS,
  payload: connected,
});

export const incrementNetworkCalls = (): IIncrementNetworkCalls => ({
  type: types.INCREMENT_NETWORK_CALLS,
});

export interface ICommonActions {
  setConnectionStatus(connected: boolean);
  incrementNetworkCalls();
}

export default {
  setConnectionStatus,
  incrementNetworkCalls,
};
