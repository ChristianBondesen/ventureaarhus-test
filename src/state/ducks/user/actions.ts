import { IFluxStandardAction } from '../FluxStandardAction';
import * as types from './types';

export interface ISetUser extends IFluxStandardAction {
  type: types.SET_USER;
  payload: types.IUser;
}

export type UserStateAction = ISetUser;

export const setUser = (user: types.IUser): ISetUser => ({
  type: types.SET_USER,
  payload: user,
});

export interface IUserActions {
  setUser(user: types.IUser);
}

export default {
  setUser,
};
