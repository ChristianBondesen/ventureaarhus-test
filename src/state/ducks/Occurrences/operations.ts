import { Dispatch } from 'redux';
import { IAppState } from '../index';
import { ReduxOperationReturnType } from '../../ReduxOperationReturnType';
import * as actions from './actions';
import loginClient from '../../../utils/loginClient';
import occurrencesClient from '../../../utils/occurrencesClient';
import { ITags } from './types';

const asyncGetOccurrences = () => {
  return async (dispatch: Dispatch<any>, getState: () => IAppState) => {
    const state = getState();

    const uri =
      'https://api.detskeriaarhus.dk/api/occurrences?startDate[after]=now'; // now is for today

    const response = await loginClient.getAsync(uri);

    if (response.ok) {
      try {
        const json = await response.json();
        const occur = json['hydra:member'];
        dispatch(actions.setOccurrenceList(occur));
      } catch (error) {
        throw error;
      }
    }
  };
};

const getTagsAsync = () => {
  return async (dispatch: Dispatch<any>, getState: () => IAppState) => {
    const state = getState();
    const uri = 'https://api.detskeriaarhus.dk/api/tags';

    const response = await occurrencesClient.getAsync(uri);
    if (response.ok) {
      try {
        const json = await response.json();
        const tags: ITags[] = json['hydra:member'];
        dispatch(actions.setOccurrencesTags(tags));
      } catch (error) {
        return error;
      }
    }
  };
};

/// Donger
const getRecommendOccurrences = () => {
  return async (dispatch: Dispatch<any>, getState: () => IAppState) => {
    const state = getState();
    const uri = 'https://api.detskeriaarhus.dk/api/occurrences';
  };
};
export type IOccurrencesOperations = {
  asyncGetOccurrences: () => ReduxOperationReturnType;
  getTagsAsync: () => ReduxOperationReturnType;
};

export default {
  asyncGetOccurrences,
  getTagsAsync,
};
