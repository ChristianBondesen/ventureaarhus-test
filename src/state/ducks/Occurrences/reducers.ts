import { OccurrenceStateAction } from './actions';
import * as types from './types';

// Interface for defualt state
export interface IOccurrenceState {
  occurrencesList: types.IOccurrence[];
  tags: types.ITags[];
}

// Default state
export const defaultState: IOccurrenceState = {
  occurrencesList: [],
  tags: [],
};

// The reducer
export const occurrencesReducer = (
  state: IOccurrenceState = defaultState,
  action: OccurrenceStateAction
): IOccurrenceState => {
  switch (action.type) {
    case types.SET_OCCURRENCES: {
      return {
        ...state,
        occurrencesList: action.payload,
      };
    }
    case types.SET_OCCURRENCES_TAGS: {
      return {
        ...state,
        tags: action.payload,
      };
    }

    default:
      return state;
  }
};

export default occurrencesReducer;
