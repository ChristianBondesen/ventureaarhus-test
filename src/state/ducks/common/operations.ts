import { Dispatch } from 'redux';
import { IAppState } from '../index';
import { ReduxOperationReturnType } from '../../ReduxOperationReturnType';
const TAG = 'common operations.ts';
import * as actions from './actions';

const asyncOperation = () => {
  return async (dispatch: Dispatch<any>, getState: () => IAppState) => {
    const state = getState();
    console.log(TAG, `call echo\n`);

    const uri = 'https://httpbin.org/get';

    dispatch(actions.incrementNetworkCalls());

    // TODO: It would be a good idea to abstract the networking into a seperate class for testing and isolation
    try {
      const result = await fetch(uri);
      if (result.ok) {
        try {
          const json = await result.json();
          // TODO: Do something with JSON
          console.log(TAG, `${JSON.stringify(json)}\n`);
          return json;
        } catch (error) {
          console.log(TAG, `No JSON returned\n`);
          return;
        }
      } else {
        // TODO: Handle result!=ok
      }
    } catch (error) {
      // TODO: Handle error
    }
  };
};

export type ICommonOperations = {
  asyncOperation: () => ReduxOperationReturnType;
};

export default {
  asyncOperation,
};
