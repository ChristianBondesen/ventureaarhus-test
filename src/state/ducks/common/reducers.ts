import { CommonStateAction } from './actions';
import * as types from './types';

export interface ICommonState {
  isConnected: boolean;
  networkCalls: number;
}

const TAG = 'reducers.ts';

export const defaultState: ICommonState = {
  isConnected: true,
  networkCalls: 0,
};

export const commonReducer = (
  state: ICommonState = defaultState,
  action: CommonStateAction
): ICommonState => {
  switch (action.type) {
    case types.CHANGE_CONNECTION_STATUS:
      return {
        ...state,
        isConnected: action.payload,
      };

    case types.INCREMENT_NETWORK_CALLS:
      return {
        ...state,
        networkCalls: state.networkCalls + 1,
      };

    default:
      return state;
  }
};

export default commonReducer;
