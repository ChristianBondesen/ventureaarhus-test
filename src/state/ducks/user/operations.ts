import { Dispatch } from 'redux';
import { IAppState } from '..';
import loginClient from '../../../utils/loginClient';
import { ReduxOperationReturnType } from '../../ReduxOperationReturnType';
import * as actions from './actions';
import { NavigationActions } from 'react-navigation';
import { RouteNames } from '../../../enums/navigationEnums';

const uri = 'http://questaarhusapi.azurewebsites.net/Account/Login';

const loginAsync = (username: string, password: string) => {
  return async (dispatch: Dispatch<any>, getState: () => IAppState) => {
    const state = getState();

    const response = await loginClient.postAsync(uri, username, password);

    if (response.ok) {
      try {
        const respText = await response.text();
        dispatch(actions.setToken(respText));
        dispatch(
          NavigationActions.navigate({ routeName: RouteNames.OccurrencesNav })
        );
      } catch (error) {
        throw error;
      }
    }
  };
};

export type IUserOperations = {
  loginAsync: (username: string, password: string) => ReduxOperationReturnType;
};

export default {
  loginAsync,
};
