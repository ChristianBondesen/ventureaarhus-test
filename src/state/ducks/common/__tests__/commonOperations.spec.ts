import mockStore from 'redux-mock-store';
import * as actions from './../actions';
import * as operations from './../operations';
import { defaultState, ICommonState } from '../reducers';
import { IFluxStandardAction } from '../../FluxStandardAction';

interface InitialStateForTest {
  common: ICommonState;
}

const initial: InitialStateForTest = {
  common: defaultState,
};

describe('Common operations tests', () => {
  beforeEach(() => {
    mockStore().clearActions();
    jest.resetAllMocks();
  });

  it(`dispatches increment call from operation`, async () => {
    const expectedActions: IFluxStandardAction[] = [
      actions.incrementNetworkCalls(),
    ];

    const store = mockStore(initial);
    await store.dispatch(operations.default.asyncOperation());

    // NB: Note that you CAN NOT assert on state using this library, only on which actions has been dispatched to it
    expect(store.getActions()).toEqual(expectedActions);
  });

  it(`dispatches increment call from operation - snapshot example`, async () => {
    const store = mockStore(initial);
    await store.dispatch(operations.default.asyncOperation());
    expect(store.getActions()).toMatchSnapshot();
  });
});
