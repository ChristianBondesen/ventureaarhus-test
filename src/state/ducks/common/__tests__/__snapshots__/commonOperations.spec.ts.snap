// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common operations tests dispatches increment call from operation - snapshot example 1`] = `
Array [
  Object {
    "type": "common/INCREMENT_NETWORK_CALLS",
  },
]
`;
